lib/Optimizer/Transforms/LiftArrayAlloc.cpp

```diff
@@ -160,6 +167,7 @@ class AllocaPattern : public OpRewritePattern<cudaq::cc::AllocaOp> {
         if (!u)
           return nullptr;
         if (auto store = dyn_cast<cudaq::cc::StoreOp>(u)) {
+          storeSets[index].insert(u);
           if (op.getOperation() == store.getPtrvalue().getDefiningOp() &&
               isa_and_present<arith::ConstantOp, complex::ConstantOp>(
                   store.getValue().getDefiningOp())) {
```

Comment:
How does any of this get past the check here? This is intended to check that we do not have multiple stores to the same offset of the array.

Reply:
I ran a little test, and this check correctly blocks the transformation.

```mlir
func.func @f() {
    %0 = cc.alloca !cc.array<i64 x 1>
    %1 = arith.constant 1 : i64
    %2 = cc.compute_ptr %0[0] : (!cc.ptr<!cc.array<i64 x 1>>) -> !cc.ptr<i64>
    cc.store %1, %2 : !cc.ptr<i64>
    %3 = arith.constant 3 : i64
    cc.store %3, %2 : !cc.ptr<i64>
    return
}
```

Running lift-array-alloc does not change this code, which is the correct thing to happen.

Reply:
I think the failing example differs on the second store not using a constant to store to a pointer

Reply:
Found the reasons why this check does not detect two stores to the same pointer: 1) the second store is appearing first in the list of users of the ptr that we are processing 2) it does not store a constant to the pointer. Updated the check.